c_crypto.c

```diff
@@ -166,120 +174,365 @@ unsigned char *digest_message_SHA_256(unsigned char *message,
     if (EVP_DigestInit_ex(mdctx, EVP_sha256(), NULL) != 1) {
         print_last_error("EVP_DigestInit_ex failed");
     }
-    if (EVP_DigestUpdate(mdctx, message, message_length) != 1) {
+    if (EVP_DigestUpdate(mdctx, data, data_len) != 1) {
         print_last_error("EVP_DigestUpdate failed");
     }
-    unsigned char *digest =
-        (unsigned char *)OPENSSL_malloc(EVP_MD_size(EVP_sha256()));
-    if (EVP_DigestFinal_ex(mdctx, digest, digest_len) != 1) {
+    if (EVP_DigestFinal_ex(mdctx, md5_hash, md_len) != 1) {
         print_last_error("failed");
     }
     EVP_MD_CTX_destroy(mdctx);
-    return digest;
 }
 
-void AES_CBC_128_encrypt(unsigned char *plaintext,
-                         unsigned int plaintext_length, unsigned char *key,
-                         unsigned int key_length, unsigned char *iv,
-                         unsigned int iv_length, unsigned char *ret,
-                         unsigned int *ret_length) {
+const EVP_CIPHER *get_EVP_CIPHER(char enc_mode) {
+    if (enc_mode == AES_128_CBC) {
+        return EVP_aes_128_cbc();
+    } else if (enc_mode == AES_128_CTR) {
+        return EVP_aes_128_ctr();
+    } else if (enc_mode == AES_128_GCM) {
+        return EVP_aes_128_gcm();
+    } else {
+        error_exit("Encryption type not supported.");
+    }
+    return NULL;
+}
+
+int encrypt_AES(unsigned char *plaintext, unsigned int plaintext_length,
+                unsigned char *key, unsigned char *iv, char enc_mode,
+                unsigned char *ret, unsigned int *ret_length) {
     EVP_CIPHER_CTX *ctx = EVP_CIPHER_CTX_new();
-    EVP_EncryptInit_ex(ctx, EVP_aes_128_cbc(), NULL, key, iv);
+    if (!EVP_EncryptInit_ex(ctx, get_EVP_CIPHER(enc_mode), NULL, key, iv)) {
+        // Error
+        EVP_CIPHER_CTX_free(ctx);
+        print_last_error("EVP_EncryptInit_ex failed");
+        return 1;
+    }
+
+    if (enc_mode == AES_128_GCM) {
+        if (!EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_IVLEN,
+                                 AES_128_GCM_IV_SIZE,
+                                 NULL)) {  // Set IV length to 16 bytes
```

Comment:
Please correct the comment (number of bytes for GCM IV.)

Reply:
Fixed 27946ce62ef4801134ca26b71c1f488e3a2b0d76.